### PR TITLE
fetch-artifacts-koji: add content-generator components metadata

### PR DESCRIPTION
**Pull request by hjmodi**
add kojifile components to output metadata for components fetched
using fetch-artifacts-koji

Added a plugin to generate maven metadata for fetch-artifacts-koji. These metadata will then be fetched by exit_koji import and added to image components alongside RPM components. This will later help us find sources for such middleware components to build source containers.

Main epic - MMENG-829
Story - MMENG-1275

Signed-off-by: Harsh Modi <hmodi@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [N/A] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [N/A] New feature can be disabled from a configuration file

### CODE REVIEW

File: tests/plugins/test_generate_maven_metadata.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    if additional:
+        reactor_map.update(additional)
+
+    workflow.plugin_workspace[ReactorConfigPlugin.key] = {
+        WORKSPACE_CONF_KEY: ReactorConfig(reactor_map)
+    }
+
+
+@responses.activate  # noqa
+def test_generate_maven_metadata(tmpdir, docker_tasker, user_params):
```
Comment (MartinBasti) [inline tests/plugins/test_generate_maven_metadata.py:283]: Do you think that would be possible to utilize this mocking here?
https://github.com/containerbuildsystem/atomic-reactor/blob/master/tests/mock_env.py

Comment (hjmodi) [reply tests/plugins/test_generate_maven_metadata.py:238]: Updated to use mock_env.

File: tests/plugins/test_generate_maven_metadata.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+           .make_orchestrator()
+           .set_reactor_config(r_c_m))
+    env.workflow.source = MockSource(tmpdir)
+    mock_koji_session()
+    mock_fetch_artifacts_by_nvr(str(tmpdir))
+    mock_nvr_downloads()
+
+    runner = (env.create_runner(docker_tasker))
+
+    results = runner.run()
```
Comment (tkdchen) [inline tests/plugins/test_generate_maven_metadata.py:256]: Probably these two lines could be merged into `results = env.create_runner(docker_tasker).run()`

File: tests/plugins/test_generate_maven_metadata.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    plugin_result = results[GenerateMavenMetadataPlugin.key]
+
+    assert len(plugin_result.get('components', [])) == len(DEFAULT_ARCHIVES)
```
Comment (tkdchen) [inline tests/plugins/test_generate_maven_metadata.py:259]: Also assert the content of the generated components to ensure the correctness?

File: tests/plugins/test_generate_maven_metadata.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    for archive in archives:
+        url = pi.mavenbuild(build_info) + '/' + pi.mavenfile(archive)
+        # Use any overrides for this archive ID
+        archive_overrides = overrides.get(archive['id'], {})
+        status = archive_overrides.get('status', 200)
+        body = archive_overrides.get('body', archive['filename'] + archive['group_id'])
+        responses.add(responses.GET, url, body=body, status=status)
+
+
+@responses.activate  # noqa
```
Comment (tkdchen) [inline tests/plugins/test_generate_maven_metadata.py:237]: Is `# noqa` necessary? `responses` is used through the codebase, `# noqa` is not used for this decorator at other place.

File: tests/plugins/test_generate_maven_metadata.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+def mock_fetch_artifacts_by_nvr(tmpdir, contents=None):
+    if contents is None:
+        contents = dedent("""\
+            - nvr: com.sun.xml.bind.mvn-jaxb-parent-2.2.11.4-1
+            """)
+
+    with open(os.path.join(tmpdir, REPO_FETCH_ARTIFACTS_KOJI), 'w') as f:
+        f.write(contents)
+        f.flush()
```
Comment (tkdchen) [inline tests/plugins/test_generate_maven_metadata.py:215]: Suggest to pass the tmpdir directly and `tmpdir.join(REPO_FETCH_ARTIFACTS_KOJI).write_text(content, 'utf-8')`

File: tests/plugins/test_generate_maven_metadata.py (hunk 3 above)
Comment (tkdchen) [inline tests/plugins/test_generate_maven_metadata.py:259]: The `DEFAULT_ARCHIVES` is not empty, so it should make no sense to get a default components value `[]` from plugin result. This line could confuse.

The assertion on the empty components in the plugin result could be another test instead.

File: tests/plugins/test_generate_maven_metadata.py (hunk 2 above)
Comment (hjmodi) [reply tests/plugins/test_generate_maven_metadata.py]: Done.

File: tests/plugins/test_generate_maven_metadata.py (hunk 3 above)
Comment (hjmodi) [reply tests/plugins/test_generate_maven_metadata.py]: Done

File: tests/plugins/test_generate_maven_metadata.py (hunk 4 above)
Comment (hjmodi) [reply tests/plugins/test_generate_maven_metadata.py]: Removed it. Not sure what it does. I thought it was some sort of note to  QA.

File: tests/plugins/test_generate_maven_metadata.py (hunk 5 above)
Comment (hjmodi) [reply tests/plugins/test_generate_maven_metadata.py]: Done

File: tests/plugins/test_generate_maven_metadata.py (hunk 3 above)
Comment (hjmodi) [reply tests/plugins/test_generate_maven_metadata.py]: Updated. I believe comparing lengths should cover the empty components case as well.

File: atomic_reactor/plugins/exit_koji_import.py [hunk 6]
```diff
@@ -23,7 +23,7 @@
 from atomic_reactor.plugins.pre_check_and_set_rebuild import is_rebuild
 from atomic_reactor.util import (OSBSLogs, get_parent_image_koji_data, get_manifest_media_version,
                                  is_manifest_list)
-from atomic_reactor.utils.koji import get_buildroot as koji_get_buildroot
+from atomic_reactor.utils.koji import get_buildroot as koji_get_buildroot, get_maven_components
 from atomic_reactor.utils.koji import get_output as koji_get_output
 from atomic_reactor.utils.koji import (
```
Comment (ben-alkov) [inline atomic_reactor/plugins/exit_koji_import.py:28]: @hjmodi; Nit-picky, but I would much prefer that `get_maven_components` be imported in this block, e.g.

```python
from atomic_reactor.utils.koji import (
    generate_koji_upload_dir, add_custom_type,
    get_source_tarball_output, get_remote_source_json_output,
    get_maven_components
)
```

tl;dr `get_buildroot` and `get_output` are being aliased very purposefully, the other modules (the ones in the block) are simply straight imports.

File: atomic_reactor/utils/koji.py [hunk 7]
```diff
@@ -258,6 +296,16 @@ def get_koji_module_build(session, module_spec):
     return build, rpm_list
 
 
+def get_maven_components(workflow):
+    """
+    Get list of maven components generated by post_generate_maven_metadata
+    :param workflow: DockerBuildWorkflow instance
+    :return: list, maven components
```
Comment (ben-alkov) [inline atomic_reactor/utils/koji.py:303]: @hjmodi; We're not super-consistent about this ourselves, but we seem to moving towards e.g.

```suggestion
    # e.g. a short text explanation of what we return
    :return: A list of maven components. Each component is blah blah
    # e.g. a Python type(s)
    :rtype: list[]  # What is a maven component's type?
```

File: atomic_reactor/plugins/exit_koji_import.py (hunk 6 above)
Comment (hjmodi) [reply atomic_reactor/plugins/exit_koji_import.py:28]: Updated

File: atomic_reactor/utils/koji.py (hunk 7 above)
Comment (hjmodi) [reply atomic_reactor/utils/koji.py]: Added info.

File: atomic_reactor/utils/koji.py [hunk 8]
```diff
@@ -258,6 +296,17 @@ def get_koji_module_build(session, module_spec):
     return build, rpm_list
 
 
+def get_maven_components(workflow):
+    """
+    Get list of maven components generated by post_generate_maven_metadata
+    :param workflow: DockerBuildWorkflow instance
+    :return: a list of maven components. Each component is a kojifile btype metadata dictionary
+    :rtype: list[dict] # list of maven metadata dictionaries
```
Comment (ben-alkov) [inline atomic_reactor/utils/koji.py:304]: :rocket:

**Review by MartinBasti — APPROVED** [review]

**Comment by MartinBasti** [comment]
https://osbs.readthedocs.io/en/latest/users.html#koji-build-metadata-integration
> In the future, a reference of each artifact fetched by OSBS will be added to the koji build metadata once imported via content generator API

Please update osbs-docs, future is now \o/ .

**Comment by hjmodi** [comment]
> https://osbs.readthedocs.io/en/latest/users.html#koji-build-metadata-integration
> 
> > In the future, a reference of each artifact fetched by OSBS will be added to the koji build metadata once imported via content generator API
> 
> Please update osbs-docs, future is now \o/ .

https://github.com/containerbuildsystem/osbs-docs/pull/180 Ready for review.

File: atomic_reactor/plugins/exit_koji_import.py [hunk 9]
```diff
@@ -373,6 +374,14 @@ def add_log_type(output):
         metadata_version = 0
 
         worker_metadatas = self.workflow.postbuild_results.get(PLUGIN_FETCH_WORKER_METADATA_KEY)
+        maven_components = get_maven_components(self.workflow)
+
+        # add maven components alongside RPM components
+        for platform in worker_metadatas.keys():
+            for component in worker_metadatas[platform]['output']:
```
Comment (rcerven) [inline atomic_reactor/plugins/exit_koji_import.py:381]: this isn't component but entries in output
so call it 'output'

for output in worker_metadatas[platform]['output']:

Comment (rcerven) [inline atomic_reactor/plugins/exit_koji_import.py:381]: this isn't component but entries in output
so call it 'output'

for output in worker_metadatas[platform]['output']:

**Review by rcerven — CHANGES_REQUESTED** [review]

**Comment by rcerven** [comment]
Looks good, just that 1 comment

File: atomic_reactor/plugins/exit_koji_import.py (hunk 9 above)
Comment (hjmodi) [reply atomic_reactor/plugins/exit_koji_import.py]: Done. Named it worker output to avoid confusion with the output variable a few lines below.

**Review by rcerven — APPROVED** [review]

**Review by tkdchen — APPROVED** [review]

**Comment by rcerven** [comment]
rn updated
